homeassistant/components/velbus/__init__.py

```diff
@@ -39,64 +48,91 @@ async def async_setup(hass, config):
             DOMAIN, context={"source": SOURCE_IMPORT}, data=data
         )
     )
-
     return True
 
 
+async def velbus_connect_task(
+    controller: Velbus, hass: HomeAssistant, entry_id: str
+) -> None:
+    """Task to offload the long running connect."""
+    await controller.connect()
+
+
 async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
     """Establish connection with velbus."""
     hass.data.setdefault(DOMAIN, {})
 
-    def callback():
-        modules = controller.get_modules()
-        discovery_info = {"cntrl": controller}
-        for platform in PLATFORMS:
-            discovery_info[platform] = []
-        for module in modules:
-            for channel in range(1, module.number_of_channels() + 1):
-                for platform in PLATFORMS:
-                    if platform in module.get_categories(channel):
-                        discovery_info[platform].append(
-                            (module.get_module_address(), channel)
-                        )
-        hass.data[DOMAIN][entry.entry_id] = discovery_info
-
-        for platform in PLATFORMS:
-            hass.add_job(hass.config_entries.async_forward_entry_setup(entry, platform))
-
-    try:
-        controller = velbus.Controller(entry.data[CONF_PORT])
-        controller.scan(callback)
-    except velbus.util.VelbusException as err:
-        _LOGGER.error("An error occurred: %s", err)
-        raise ConfigEntryNotReady from err
-
-    def syn_clock(self, service=None):
-        try:
-            controller.sync_clock()
-        except velbus.util.VelbusException as err:
-            _LOGGER.error("An error occurred: %s", err)
-
-    hass.services.async_register(DOMAIN, "sync_clock", syn_clock, schema=vol.Schema({}))
-
-    def set_memo_text(service):
+    controller = Velbus(entry.data[CONF_PORT])
+    hass.data[DOMAIN][entry.entry_id] = {}
+    hass.data[DOMAIN][entry.entry_id]["cntrl"] = controller
+    hass.data[DOMAIN][entry.entry_id]["tsk"] = hass.async_create_task(
+        velbus_connect_task(controller, hass, entry.entry_id)
+    )
+
+    hass.config_entries.async_setup_platforms(entry, PLATFORMS)
+
+    if hass.services.has_service(DOMAIN, SERVICE_SCAN):
+        return True
+
+    def get_entry_id(interface: str) -> str | None:
+        for entry in hass.config_entries.async_entries(DOMAIN):
+            if "port" in entry.data and entry.data["port"] == interface:
+                return entry.entry_id
+        _LOGGER.warning("Can not find the config entry for: %s", interface)
+        return None
+
+    def check_entry_id(interface: str):
+        for entry in hass.config_entries.async_entries(DOMAIN):
+            if "port" in entry.data and entry.data["port"] == interface:
+                return interface
```

Comment:
If we return the config entry id here we don't need to look up the config entry id in the service handler with `get_entry_id`.